Std/Classes/LawfulAlternative.lean

```diff
@@ -0,0 +1,73 @@
+import Std.Classes.LawfulMonad
+import Lean.Linter
+
+/-- Combination class of `Monad` and `Alternative` with a common `Applicative` component -/
+class MonadAlternative (m : Type _ → Type _) extends Monad m, Alternative m
+
+-- Missing docs in core
+attribute [nolint docBlame] MonadAlternative.failure MonadAlternative.orElse
+
+/-- Basic laws for `Alternative` -/
+class LawfulAlternative (f : Type _ → Type _) [Alternative f] extends LawfulApplicative f : Prop where
+  /-- Map failure is failure -/
+  map_failure (g : α → β) : g <$> (failure : f α) = failure
+  /-- Map distributes over `orElse` -/
+  map_orElse (g : α → b) (x y : f α) : g <$> (x <|> y) = (g <$> x <|> g <$> y)
+  /-- Failure is a left identity for `orElse`  -/
+  failure_orElse (x : f α) : (failure <|> x) = x
+  /-- Failure is a right identity for `orElse` -/
+  orElse_failure (x : f α) : (x <|> failure) = x
+  /-- `orElse` is associative -/
+  orElse_assoc (x y z : f α) : (x <|> (y <|> z)) = ((x <|> y) <|> z)
+  /-- `pure` always succeeds -/
+  pure_orElse (x : α) (y : f α) : (pure x <|> y) = pure x
```

Comment:
I've had occasion to want `Alternative List`, with operations `failure := []`, `<|> := ++` and `pure a := [a]`. This would satisfy all these axioms except for `pure_orElse`. 

There seems to be some debate in the Haskell community about the laws for `Alternative`: [1](https://en.wikibooks.org/wiki/Haskell/Alternative_and_MonadPlus#Alternative_and_MonadPlus_laws) [2](https://wiki.haskell.org/MonadPlus_reform_proposal).

I'd certainly like `List` to be `LawfulAlternative`!